### DISCUSSION TITLE
how to clip or hide the part of object inside a group?

### DISCUSSION THREAD

**Discussion by wuxingping** [fabricjs/fabric.js#8647] [General]
First, the group turn layout to fixed and the version is v6, i want to clip or hide these content when the one inside group drag exceed the scope of group. but it still exist when i drag to inside group
<img width="489" alt="1675401670564" src="https://user-images.githubusercontent.com/41353266/216518967-c8b9a167-d8d5-45fd-87f2-951acf976cb2.png">
![1675401484971](https://user-images.githubusercontent.com/41353266/216518772-46394cdd-415d-48b2-963c-ef3ead7aceb4.jpg)

**Comment by iocoker** [comment]
Set the layout property to `clip-path` instead.
I'm assuming the black part is a `clipPath` Rect?

**Reply by wuxingping** [reply]
the black part is group,the group turn backgroundColor to black.

**Reply by iocoker** [reply]
You will need to have a `clipPath` to define the group boundary, perhaps a Rect.
See the Pink circle in this fiddle: http://jsfiddle.net/Lxjegow3/8/

**Reply by wuxingping** [reply]
i try.but it wrong

**Reply by wuxingping** [reply]
> You will need to have a `clipPath` to define the group boundary, perhaps a Rect. See the Pink circle in this fiddle: http://jsfiddle.net/Lxjegow3/8/

hi, i am in China,i couldn't open this link.i coded as you say but it doesnot work.
`   var rect = new fabric.Rect({
            left:120,
            width:400,
            height:400,
            fill:"red",
            strokeWidth:10,
            stroke:"pink",
    })
        var group = new fabric.Group([], {
            subTargetCheck: true,
            left:120,
            width:400,
            height:400,
            top:0,
            fill:"red",  
            rx:10,
            ry:10,      
            layout:"clip-path",            
            clipPath:rect,      
            interactive: true
        })`

**Reply by ShaMan123** [reply]
show what isn't working

**Reply by wuxingping** [reply]
> show what isn't working
on the hand, this is  wrong situation.when i turn preserveObjectStacking false.
![1675840512995](https://user-images.githubusercontent.com/41353266/217460310-61157a64-e405-4b28-8454-2d91d8e6c34c.jpg)
on other hand, this also  is  wrong situation.when i turn preserveObjectStacking true.
![1675840732635](https://user-images.githubusercontent.com/41353266/217460915-4578f514-8a5b-4a80-8148-d75bfa0120bb.jpg)

**Reply by wuxingping** [reply]
![image](https://user-images.githubusercontent.com/41353266/217462798-18df9d5e-b707-41ca-9a98-ae8d52bc4d59.png)

**Reply by ShaMan123** [reply]
I believe the problem is not in your code but in the version of fabric used in the 2 cases.
That example use the `v6-alpha` branch I believe which includes #7882 which is what fixed this. But **the branch is stale**.
#8665 is a port to master but rendering and caching issues are not resolved yet as they are in `v6-alpha`.

**Reply by wuxingping** [reply]
> I believe the problem is not in your code but in the version of fabric used in the 2 cases. That example use the `v6-alpha` branch I believe which includes #7882 which is what fixed this. But **the branch is stale**. #8665 is a port to master but rendering and caching issues are not resolved yet as they are in `v6-alpha`.

so what's the useful version about that? master branch? the way of download is that  npm i --save fabricjs/fabric.js#master && cd ./node_modules/fabric && npm run build. is it right?

**Reply by ShaMan123** [reply]
yes, but now that beta is out you can work on a stable version `npm i --save fabric@beta` but that won't have the fix.
First of all try the `v6-alpha`, see if you manage to build (might be breaking changes between it and master, I don't remember)

Or try #8665 
`npm i --save fabricjs/fabric.js#v6-selection && cd ./node_modules/fabric && npm run build`

But still caching will be wrong so you will need to port the caching logic from #7882

**Reply by wuxingping** [reply]
> yes, but now that beta is out you can work on a stable version `npm i --save fabric@beta` but that won't have the fix. First of all try the `v6-alpha`, see if you manage to build (might be breaking changes between it and master, I don't remember)
> 
> Or try #8665 `npm i --save fabricjs/fabric.js#v6-selection && cd ./node_modules/fabric && npm run build`
> 
> But still caching will be wrong so you will need to port the caching logic from #7882

whats the branch of the right picture as i showed?

**Reply by wuxingping** [reply]
there is a trouble that i have to turn originX to center, otherwise i  can not set it in canvas,but the group is not at suitable position,as follows:
 
![1675931004475](https://user-images.githubusercontent.com/41353266/217756681-f330512f-13dd-4df3-a3f7-dc64e38c45a2.jpg)

`  var rect = new Rect({
        width: 500,
        height:500,
        originX: "center",
        originY: "center",         
      });
      var  shape = new Container([
new Rect({
        width: 100,
        height:100,
       fill:"red"
      }),new Rect({
      width: 200,
        height:200,
       fill:"pink"
      })
      ],
        {
          subTargetCheck:true,
          interactive:true,
          layout:"clip-path",  
          clipPath:rect,
        }
      )`

**Reply by ShaMan123** [reply]
> whats the branch of the right picture as i showed?

v6-alpha

> there is a trouble that i have to turn originX to center, otherwise i can not set it in canvas,but the group is not at suitable position,as follows:

It all depends on caching I think
But this is the most recent fabric has to offer, the bleeding edge so you may cut yourself 😉
